platform/remote/dev/phidget-sensor.c

```diff
@@ -0,0 +1,138 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/*---------------------------------------------------------------------------*/
+/**
+ * \addtogroup remote-phidget-sensor
+ * @{
+ *
+ * \file
+ * Generic driver for the Re-Mote Phidget/ADC sensors
+ */
+/*---------------------------------------------------------------------------*/
+#include "contiki.h"
+#include "sys/clock.h"
+#include "dev/ioc.h"
+#include "dev/gpio.h"
+#include "dev/adc.h"
+#include "dev/phidget-sensor.h"
+#include "dev/remote-sensors.h"
+
+#include <stdint.h>
+/*---------------------------------------------------------------------------*/
+#define ADC_PHIDGET_PORT_BASE      GPIO_PORT_TO_BASE(ADC_PHIDGET_PORT)
+#define ADC_PHIDGET_ADC2_PIN_MASK  GPIO_PIN_MASK(ADC_PHIDGET_ADC2_PIN)
+#define ADC_PHIDGET_ADC3_PIN_MASK  GPIO_PIN_MASK(ADC_PHIDGET_ADC3_PIN)
+/*---------------------------------------------------------------------------*/
+static uint8_t decimation_rate;
+static uint8_t shift;
+/*---------------------------------------------------------------------------*/
+static int
+set_decimation_rate(uint8_t rate)
+{
+  switch(rate) {
+  case SOC_ADC_ADCCON_DIV_64:
+    decimation_rate = rate;
+    shift = SOC_ADC_7_BIT_RSHIFT;
+    break;
+  case SOC_ADC_ADCCON_DIV_128:
+    decimation_rate = rate;
+    shift = SOC_ADC_9_BIT_RSHIFT;
+    break;
+  case SOC_ADC_ADCCON_DIV_256:
+    decimation_rate = rate;
+    shift = SOC_ADC_10_BIT_RSHIFT;
+    break;
+  case SOC_ADC_ADCCON_DIV_512:
+    decimation_rate = rate;
+    shift = SOC_ADC_12_BIT_RSHIFT;
+    break;
+  default:
+    return REMOTE_SENSORS_ERROR;
+  }
+
+  return decimation_rate;
+}
+/*---------------------------------------------------------------------------*/
+static int
+value(int type)
+{
+  uint8_t channel;
+  int16_t res;
+
+  switch(type) {
+  case PHIDGET_SENSORS_ADC2:
+    channel = SOC_ADC_ADCCON_CH_AIN0 + ADC_PHIDGET_ADC2_PIN;
+    break;
+  case PHIDGET_SENSORS_ADC3:
+    channel = SOC_ADC_ADCCON_CH_AIN0 + ADC_PHIDGET_ADC3_PIN;
+    break;
+  default:
+    return REMOTE_SENSORS_ERROR;
+  }
+
+  res = adc_get(channel, SOC_ADC_ADCCON_REF_INT, decimation_rate);
+
+  /* Shift right by a number of bits depending on the decimation rate */
+  return res >> shift;
```

Comment:
It would be better to remove all the ADC shift stuff from this PR, as discussed in #968.